src/abstracts/SablierV2Lockup.sol

```diff
@@ -389,26 +518,135 @@ abstract contract SablierV2Lockup is
+        }
+
+        // Calculate the sender's and the recipient's amount.
+        uint128 senderAmount = amounts.deposited - streamedAmount;
+        uint128 recipientAmount = streamedAmount - amounts.withdrawn;
+
+        // Effects: mark the stream as canceled.
+        _streams[streamId].wasCanceled = true;
+
+        // Effects: make the stream not cancelable anymore, because a stream can only be canceled once.
+        _streams[streamId].isCancelable = false;
+
+        // Effects: If there are no assets left for the recipient to withdraw, mark the stream as depleted.
+        if (recipientAmount == 0) {
+            _streams[streamId].isDepleted = true;
+        }
+
+        // Effects: set the refunded amount.
+        _streams[streamId].amounts.refunded = senderAmount;
+
+        // Retrieve the sender and the recipient from storage.
+        address sender = _streams[streamId].sender;
+        address recipient = _ownerOf(streamId);
+
+        // Retrieve the ERC-20 asset from storage.
+        IERC20 asset = _streams[streamId].asset;
+
+        // Interactions: refund the sender.
+        asset.safeTransfer({ to: sender, value: senderAmount });
+
+        // Log the cancellation.
+        emit ISablierV2Lockup.CancelLockupStream(streamId, sender, recipient, asset, senderAmount, recipientAmount);
+
+        // Emits an ERC-4906 event to trigger an update of the NFT metadata.
+        emit MetadataUpdate({ _tokenId: streamId });
+
+        // Interactions: if the recipient is a contract, try to invoke the cancel hook on the recipient without
+        // reverting if the hook is not implemented, and without bubbling up any potential revert.
+        if (recipient.code.length > 0) {
+            try ISablierV2Recipient(recipient).onLockupStreamCanceled({
+                streamId: streamId,
+                sender: sender,
+                senderAmount: senderAmount,
+                recipientAmount: recipientAmount
+            }) { } catch { }
+        }
+    }
 
     /// @dev See the documentation for the user-facing functions that call this internal function.
-    function _renounce(uint256 streamId) internal virtual;
+    function _renounce(uint256 streamId) internal {
+        // Checks: the stream is cancelable.
+        if (!_streams[streamId].isCancelable) {
+            revert Errors.SablierV2Lockup_StreamNotCancelable(streamId);
+        }
+
+        // Effects: renounce the stream by making it not cancelable.
+        _streams[streamId].isCancelable = false;
+    }
 
     /// @dev See the documentation for the user-facing functions that call this internal function.
-    function _withdraw(uint256 streamId, address to, uint128 amount) internal virtual;
+    function _withdraw(uint256 streamId, address to, uint128 amount) internal {
+        // Effects: update the withdrawn amount.
+        _streams[streamId].amounts.withdrawn = _streams[streamId].amounts.withdrawn + amount;
```

Comment:
```suggestion
        _streams[streamId].amounts.withdrawn += amount;
```

for code readability.

Also considering how many checks we already have in `withdraw`, we can safely place it in unchecked block.

Reply:
it is actually more gas efficient to leave it as it is (small difference) despite the readability

see also this audit G-02: https://github.com/sablier-labs/audits/blob/main/v2-core/hake-2023-03-31.pdf 

> we can safely place it in unchecked block

[as said above](https://github.com/sablier-labs/v2-core/pull/813#discussion_r1486240939), let's not use `unchecked` on state change

Reply:
it is worth noting that that optimization might no longer be applicable with via IR. But let's keep it just in case.